Boost: Wire LCP Module State to Boost Settings UI

LiamSarsfield opened the pull request:
> [!NOTE]
> Merge this PR after https://github.com/Automattic/jetpack/pull/43058 is merged.

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes HOG-52 - Update LCP UI to use stored state
Fixes HOG-36

## Proposed changes:
* Implement UI components showing optimization status and controls
* Set up frontend data storage layer for optimization results

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pc9hqz-3ul-p2

## Does this pull request change what data or activity we track or use?
Yes - adds anonymous tracking events for LCP optimization actions (when users enable, disable, or trigger optimization).

## Testing instructions:
Note: This can currently only be tested locally, as [Hydra: Find the LCP element and return it #669](https://github.com/Automattic/boost-cloud/pull/669) is not deployed to the Cloud yet
* Ensure latest version of boost-developer is also checked out to ensure [Update LCP analysis requests to Boost Cloud #36](https://github.com/Automattic/boost-developer/pull/36) is applied
* Make sure to get the cloud up and running correctly. Sometimes there's a problem with the node_modules from the root of the project polluting the docker instance and causing problems. Just to be on the safe side, delete:
./boost-shield-api/node_modules
./boost-hydra/node_modules
./node_modules 
* Add `define('JETPACK_BOOST_ALPHA_FEATURES', true)`; to your wp-config.php
  * This enables access to experimental features like LCP optimization
* Enable Jetpack Boost and ensure you have at least one Cornerstone page configured
* Navigate to the Boost dashboard and verify the LCP module is visible
* Enable the LCP module and observe the UI updates to "pending" state
* Click the "Optimize" button and verify it triggers the optimization process
* Check that the status updates appropriately when optimization completes
* Verify time display works correctly after optimization finishes
* Test error handling by temporarily modifying code to force an error state

LiamSarsfield committed:
Enhance LCP optimization module with new classes and methods

- Introduced LCP_Optimizer and LCP_State classes to manage the LCP optimization process.
- Added methods for starting the optimization, managing page states, and handling errors.
- Updated Cornerstone_Provider to include a new method for generating provider keys.

LiamSarsfield committed:
changelog

dilirity committed:
Rename optimizer to analyzer

dilirity committed:
Add endpoint to store data in LCP storage

dilirity committed:
add changelog

LiamSarsfield committed:
Merge branch 'trunk' into add/boost/lcp-state-storage

LiamSarsfield committed:
Removing unnecessary functions for current MVP

dilirity committed:
Fix static analysis

dilirity committed:
Remove debug state

dilirity committed:
Merge branch 'add/boost/lcp-state-storage' into add/boost/lcp-save-to-storage

LiamSarsfield committed:
Changed the interface from Changes_Output_On_Activation to Changes_Output_After_Activation in the Lcp class.

LiamSarsfield committed:
Merge branch 'add/boost/lcp-state-storage' into add/boost/lcp-wire-storage-to-application

LiamSarsfield committed:
Refactor LCP optimization to utilize Cornerstone Provider for current request data retrieval.
Updated LCP class to optimize viewports based on LCP data, removing the previous method for fetching LCP image tags.

LiamSarsfield committed:
Add is_current_page_cornerstone method to Cornerstone_Utils and integrate it into LCP class for conditional processing based on cornerstone pages.

LiamSarsfield committed:
Enhance LCP optimization by adding defensive checks for empty HTML and adjusting the LCP HTML string before processing. This ensures more reliable tag optimization and early returns when necessary.

LiamSarsfield committed:
Adding Changelog for LCP Wire Storage

LiamSarsfield committed:
Merge remote-tracking branch 'origin/add/boost/lcp-state-storage' into add/boost/lcp-save-to-storage

LiamSarsfield committed:
Enhance LCP_State class with state management methods and validation. Introduced get_state method for retrieving and validating LCP state, ensuring consistent state handling across class methods.

LiamSarsfield committed:
Merge branch 'add/boost/lcp-save-to-storage' into add/boost/lcp-wire-storage-to-application

LiamSarsfield committed:
Implement LCP optimization state management and UI updates.

LiamSarsfield committed:
Adding Changelog for LCP Wire UI

LiamSarsfield committed:
Update LCP status messages to use translation functions for better localization support.

LiamSarsfield commented on projects/plugins/boost/app/assets/src/js/features/lcp/lcp.tsx:
```diff
@@ -3,33 +3,108 @@ import styles from './lcp.module.scss';
+
+	if ( lcpState?.status === 'pending' ) {
+		return (
+			<div className={ styles?.generating }>
```
I noticed the text colour when the UI is in the pending state is slightly different for Cloud CSS.
It seems the Cloud CSS pending text here doesn't have the generating style as it's being added as an `overrideText`
https://github.com/Automattic/jetpack/blob/813cffd39add9cb685868587916c1b3185c15883/projects/plugins/boost/app/assets/src/js/features/critical-css/status/status.tsx#L60
![Screenshot 2025-04-14 at 15 47 35](https://github.com/user-attachments/assets/bbc62e99-0eea-4b12-b6e7-d4433f51be0f)

dilirity replied:
I copied the styles almost as is from Critical CSS status. They are probably missing some selectors.

dilirity replied:
Added a ticket so we can address this in HOG-72.

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/lcp-wire-ui-state

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/lcp.tsx:
```diff
@@ -3,33 +3,108 @@ import styles from './lcp.module.scss';
+	const optimizeAction = useOptimizeLcpAction();
+
+	const handleEnable = () => {
+		recordBoostEvent( 'lcp_optimize_enabled', {} );
```
Good thinking adding tracking events, however we already have events on PHP level here:

https://github.com/Automattic/jetpack/blob/3f3451f3d2ed369c42195e18aa33aa0a88b354e7/projects/plugins/boost/app/lib/class-status.php#L94-L102

It gets called when a module is toggled:

https://github.com/Automattic/jetpack/blob/3f3451f3d2ed369c42195e18aa33aa0a88b354e7/projects/plugins/boost/app/lib/class-status.php#L57-L60

https://github.com/Automattic/jetpack/blob/21fd2c77fb3fd29b8d2c06c86dd03aee986c8f07/projects/plugins/boost/app/modules/class-modules-setup.php#L236

https://github.com/Automattic/jetpack/blob/3f3451f3d2ed369c42195e18aa33aa0a88b354e7/projects/plugins/boost/app/modules/class-modules-setup.php#L188

https://github.com/Automattic/jetpack/blob/3f3451f3d2ed369c42195e18aa33aa0a88b354e7/projects/plugins/boost/app/data-sync/class-modules-state-entry.php#L52

LiamSarsfield replied:
Good to know! I'll remove these from the frontend Boost Events. Should we still keep the lcp_optimize_clicked? Because we wouldn't know if the starting of the optimizaton happened from the toggle, or the Optimize button.

dilirity replied:
Thanks!

I think we can keep the event when a button is clicked.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/lcp.tsx:
```diff
@@ -3,33 +3,108 @@ import styles from './lcp.module.scss';
+		query.refetch();
+	};
+
+	const handleDisable = () => {
+		recordBoostEvent( 'lcp_optimize_disabled', {} );
+	};
```
Not necessary. See above comment.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/lib/stores/lcp-state-types.ts:
```diff
@@ -0,0 +1,52 @@
+import { JSONSchema } from '$lib/utils/json-types';
+import z from 'zod';
+
+// @TODO: We don't send this back from the API, but it's here for if we do.
+export const LcpErrorType = z.enum( [
+	'UrlError',
+	'HttpError',
+	'UnknownError',
+	'TimeoutError',
+	'OptimizationError',
+] );
```
Good thinking.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/lib/stores/lcp-state-types.ts:
```diff
@@ -0,0 +1,52 @@
+ * Infer Zod Types
+ */
+export type LcpState = z.infer< typeof LcpStateSchema >;
+export type Lcp_Error_Type = z.infer< typeof LcpErrorType >;
```
I_noticed_this_type_uses_different_case 😅

LiamSarsfield replied:
Not sure how this happened 🤔 Thanks for noticing.

dilirity replied:
No_Problem

dilirity commented on projects/plugins/boost/app/modules/optimizations/lcp/class-optimize-lcp-endpoint.php:
```diff
@@ -0,0 +1,40 @@
+	 * @param \WP_REST_Request $_request The request object.
+	 */
+	public function handle( $_data, $_request ) {
+		try {
```
I'm not sure if the try catch block is necessary. I can't see anything that can throw an exception.

Maybe simplify this and we can revisit if necessary when we're handling error reporting?

LiamSarsfield replied:
That's fair.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/lcp/lib/stores/lcp-state.ts:
```diff
@@ -0,0 +1,72 @@
+			onResult: ( result, _state ): LcpState => {
+				if ( result.success ) {
+					if ( callback ) {
+						callback();
```
Nitpick: I know nothing is using this code at the moment, but it wouldn't hurt if we pass the state here.

We already do this for critical css.

LiamSarsfield replied:
Makes sense! I've also updated it to be more consistent to critical css by renaming it to onSuccess

dilirity reviewed (changes requested):
Awesome work on this! 🥳 

I've added a few notes below.

LiamSarsfield committed:
Refactor LCP optimization handling and improve type definitions

- Removed unnecessary frontend event recording for enabling/disabling LCP optimization in the LCP component.
- Updated the `useOptimizeLcpAction` hook to use a more descriptive parameter name and pass the resulting state to the success callback.
- Fixed type definition for `LcpErrorType` in lcp-state-types.ts for consistency.
- Simplified the error handling in the Optimize_LCP_Endpoint class by removing unnecessary try-catch structure.

dilirity reviewed (approved):
:shipit:
